docs/Overview.md

```diff
@@ -1,96 +1,246 @@
-// A query is represented as a directed acyclic graph (DAG).
+A query defines a unit of work for the engine. It comprises one or more pipelines.
+It contains its representation, its results, and execution statistics.
+
+```go
 type Query interface {
-    Nodes() []QueryNode
-    Edges() []Edge
+    Results() Results
+    Stats() map[string]interface{}
+    Representation() Representation
+}
+
+type Representation interface {
+    Visitor // a representation can be visited for compilation/execution.
+    Lang() Language // a representation makes sense in a certain language.
+    String() string // a visual representation of the query.
+}
+
+// The results for a Query.
+type Results interface {
+    ForEach(func (Result))
 }
 
-// QueryNode is a single step in the DAG of a query
-type QueryNode interface {
-    ID() NodeID
-    // More details about what the node does
+type Result interface {
+    Name() string
+    // ... methods to iterate other the contents of the result
 }
+```
 
-// NodeID uniquely identifies a node.
-type NodeID
+The executor executes a query expressed in some language and fills its results, provided an execution context.
+There can be multiple executors, one for each language supported by the engine.
+The engine is responsible for providing the executor with an
+execution context that embeds everything that is necessary for the execution of the query, included the registered values,
+sources and transformations. 
 
-// Edge establishes a parent child relationship between two nodes.
-type Edge interface {
-    Parent() NodeID
-    Child() NodeID
+```go
+type Executor interface {
+    Execute(Query, ExecutionContext) error // fills query results.
+    Type() Language
 }
 
-// Planner computes a plan from a query and available storage interfaces
-type Planner interface {
-    Plan(Query, []Storage) Plan
+// The ExecutionContext contains any information necessary for properly executing a query.
+type ExecutionContext interface {
+    Bindings() Scope
+    GetSource(id string) Source
+    GetTransformation(id string) Transformation
+    
+    // ... examples of other components of the execution context.
+    Context() context.Context
+    MemoryAllocator() Allocator
+    Logger() Logger
 }
 
-// Plan is a DAG of the specific steps to execute.
-type Plan interface  {
-    Nodes() []PlanNode
-    Edges() []Edge
+// Scope is a set of bindings from identifiers to values.
+type Scope = map[string]Value
+```
+   
+A compiler translates a query representation in a language into another target language.
+
+```go
+type Compiler interface {
+    Compile(Representation) (Representation, err)
+    Type() CompilerType
 }
+```
 
-// PlanNode is a single step in the plan DAG.
-type PlanNode interface {
-    ID() NodeID
-    Predicates() []Predicate
+As an example, let `e` be an engine that supports `String`, `AST`, and `SemanticGraph` queries:
+
+```go
+func (e *Engine) Query(r Representation) (Query, error) {
```

Comment:
I think it would be useful to also include the compilers in here too since they transform representations to other representations. I think it would be helpful to add a mechanism to the engine to register compilers and executors.

Reply:
At first I did it, but, by making executors compile and delegate to others, it seemed useless because they would always need access to the engine for getting a compiler. We'll discuss on this.